cc/concurrent_selfplay.cc

```diff
@@ -0,0 +1,873 @@
+  if (FLAGS_cache_size_mb > 0) {
+    auto capacity =
+        BasicInferenceCache::CalculateCapacity(FLAGS_cache_size_mb);
+    MG_LOG(INFO) << "Will cache up to " << capacity
+                 << " inferences, using roughly " << FLAGS_cache_size_mb
+                 << "MB.\n";
+    inference_cache =
+        std::make_shared<ThreadSafeInferenceCache>(capacity, FLAGS_cache_shards);
+  } else {
+    inference_cache = std::make_shared<NullInferenceCache>();
+  }
+
+  // Initialize the selfplay threads.
+  std::vector<SelfplayThread> selfplay_threads;
+  std::vector<std::unique_ptr<ModelFactory>> model_factories;
+  std::string model_name;
+  {
+    absl::MutexLock lock(&mutex_);
+    auto num_threads = devices_.size() * FLAGS_threads_per_device;
+    selfplay_threads.reserve(num_threads);
+    int thread_id = 0;
+    for (const auto& device : devices_) {
+      model_factories.push_back(NewModelFactory(FLAGS_engine, device));
+      for (int i = 0; i < FLAGS_threads_per_device; ++i) {
+        // TODO(tommadams): add a method to the model factory to create multiple
+        // model instances from the same file.
+        auto model = model_factories.back()->NewModel(FLAGS_model);
+        if (model_name.empty()) {
+          model_name = model->name();
+        }
+        selfplay_threads.emplace_back(thread_id++, std::move(model), this, inference_cache);
+      }
+    }
+  }
+
+  // Start the output thread.
+  OutputThread output_thread(&output_queue_);
+  output_thread.Start();
+
+  // Start the selfplay threads.
+  for (auto& t : selfplay_threads) {
+    t.Start();
+  }
+
+  // Wait for the selfplay threads to finish.
+  for (auto& t : selfplay_threads) {
+    t.Join();
+  }
+
+  // Stop the output thread.
+  output_queue_.Push(nullptr);
+  output_thread.Join();
+  MG_CHECK(output_queue_.empty());
+
+  {
+    absl::MutexLock lock(&mutex_);
+    MG_LOG(INFO) << FormatWinStatsTable({{model_name, win_stats_}});
+  }
+}
+
+std::unique_ptr<SelfplayGame> Selfplayer::StartNewGame(
+    const std::string& model_name, bool verbose) {
+  WTF_SCOPE0("StartNewGame");
+
+  Game::Options game_options;
+  MctsTree::Options tree_options;
+  SelfplayGame::Options selfplay_options;
+
+  {
+    absl::MutexLock lock(&mutex_);
+    if (!FLAGS_run_forever && num_games_remaining_ == 0) {
+      return nullptr;
+    }
+    if (!FLAGS_run_forever) {
+      num_games_remaining_ -= 1;
+    }
+
+    game_options = game_options_;
+    game_options.resign_enabled = rnd_() >= FLAGS_disable_resign_pct;
+
+    tree_options = tree_options_;
+
+    selfplay_options.num_virtual_losses = FLAGS_virtual_losses;
+    selfplay_options.num_readouts = FLAGS_num_readouts;
+    selfplay_options.fastplay_readouts = FLAGS_fastplay_readouts;
+    selfplay_options.fastplay_frequency = FLAGS_fastplay_frequency;
+    selfplay_options.noise_mix = FLAGS_noise_mix;
+    selfplay_options.dirichlet_alpha = FLAGS_dirichlet_alpha * 361.0 / (kN * kN);
```

Comment:
if this is getting rescaled here, then dirichlet alpha should be 0.03 in the flags.  My preference would be to not transform the flag value at all.

Reply:
done